common/retval.h

```diff
@@ -35,14 +37,17 @@ T failure_value() { return 0; }
 template<typename T>
 struct retval : public retval_base {
     T _val;
-    retval(T x) : _val(x) { }
-    retval(int _errno, T val) : retval_base{(uint64_t)_errno}, _val(val) {
+    retval(const retval&) = default;
+    retval(int _errno, T val) : retval_base{_errno}, _val(val) { }
+    retval(int _errno) : retval_base{_errno}, _val(failure_value<T>()) {
         assert(failed());
     }
     retval(const retval_base& rvb) : retval_base(rvb) {
         assert(failed());
         _val = failure_value<T>();
     }
+    template<typename P, typename Q = typename std::enable_if<!std::is_integral<P>::value>::type>
```

Comment:
prevent direct construction from T, if T is an integral type, so as to avoid conflicts with error number of int.